Test/comp/Variance.dfy

```diff
@@ -0,0 +1,190 @@
+    const x : int;
+    const y : seq<T>;
+
+    function method A(x : T) : int { 0 }
+    static function method sA(x : T) : int { 0 }
+    function method gA(ghost x : T) : int { 0 }
+    function method B(x : seq<T>) : int { 0 }
+    function method C(x : int) : seq<T> { [] }
+    function method D(x : T) : T { x }
+
+    method mA(x : T) returns (r: int) { r := 0; }
+    method mD(x : T) returns (r: T) { r := x; }
+}
+
+datatype In<T> = In(T){
+    const x : int;
+    const y : seq<T>;
+
+    function method A(x : T) : int { 0 }
+    static function method sA(x : T) : int { 0 }
+    function method gA(ghost x : T) : int { 0 }
+    function method B(x : seq<T>) : int { 0 }
+    function method C(x : int) : seq<T> { [] }
+    function method D(x : T) : T { x }
+
+    method mA(x : T) returns (r: int) { r := 0; }
+    method mD(x : T) returns (r: T) { r := x; }
+}
+
+datatype Con<-T> = Con {
+    const x : int;
+    const y : seq<T>;
+
+    function method A(x : T) : int { 0 }
+    static function method sA(x : T) : int { 0 }
+    function method gA(ghost x : T) : int { 0 }
+    function method B(x : seq<T>) : int { 0 }
+    function method C(x : int) : seq<T> { [] }
+    function method D(x : T) : T { x }
+
+    method mA(x : T) returns (r: int) { r := 0; }
+    method mD(x : T) returns (r: T) { r := x; }
+}
+
+codatatype CCo<+T> = CCo(T){
+    const x : int;
+    const y : seq<T>;
+
+    function method A(x : T) : int { 0 }
+    static function method sA(x : T) : int { 0 }
+    function method gA(ghost x : T) : int { 0 }
+    function method B(x : seq<T>) : int { 0 }
+    function method C(x : int) : seq<T> { [] }
+    function method D(x : T) : T { x }
+
+    method mA(x : T) returns (r: int) { r := 0; }
+    method mD(x : T) returns (r: T) { r := x; }
+}
+
+codatatype CIn<T> = CIn(T){
+    const x : int;
+    const y : seq<T>;
+
+    function method A(x : T) : int { 0 }
+    static function method sA(x : T) : int { 0 }
+    function method gA(ghost x : T) : int { 0 }
+    function method B(x : seq<T>) : int { 0 }
+    function method C(x : int) : seq<T> { [] }
+    function method D(x : T) : T { x }
+
+    method mA(x : T) returns (r: int) { r := 0; }
+    method mD(x : T) returns (r: T) { r := x; }
+}
+
+codatatype CCon<-T> = CCon {
+    const x : int;
+    const y : seq<T>;
+
+    function method A(x : T) : int { 0 }
+    static function method sA(x : T) : int { 0 }
+    function method gA(ghost x : T) : int { 0 }
+    function method B(x : seq<T>) : int { 0 }
+    function method C(x : int) : seq<T> { [] }
+    function method D(x : T) : T { x }
+
+    method mA(x : T) returns (r: int) { r := 0; }
+    method mD(x : T) returns (r: T) { r := x; }
+}
+
+trait X {}
+
+class Int extends X {
+  constructor Int() { }
+}
+
+method Covariant() {
+  var i := new Int.Int();
+  var a: Co<Int> := Co(i);
+  var b: Co<X>; // compilation error: compilation does not support trait types as a type parameter; consider introducing a ghost
```

Comment:
Can you mention that it's only Java compilation that's giving the error?

Reply:
I do have something to that effect at the top.

Reply:
I saw that, but I still got confused